### DISCUSSION TITLE
Rust 课程大纲升级 （阶段一）

### DISCUSSION THREAD

**Discussion by wubx** [wubx/rust-in-databend#17] [General]
不知不觉 Rust 培养提高计划进行了 23 期，目前课程给大家的感觉还是内容太散。 目前计划把课程做一个升级，让内容更加系统一点。 但这里不是一个 0 基础的学习教程（例如，什么是变量， 如何写一个流程控制，这些内容需要大家自行阅读一下 Rust 相关教程，可能 2-3 天就可以搞定） ，我们希望这个教程定位在，帮你在个人学习中遇到的困难或是想不明白地方，在这个教程中给完善起来。  

## 课程同步地址：
- 腾讯会议：  https://meeting.tencent.com/dm/mlhs6YsYuFRb  会议 ID： 258 663 900
- B站： http://live.bilibili.com/22123086
- PPT发布： https://github.com/wubx/rust-in-databend
- Databend公众号:  Databend 
- Rust中文社区 : rustcc.cn 

目前课程分为 ： 周四晚上，周日晚上，其中周四晚上不是一个固定时间（课程内容属于邀请制）， 周日晚上的内容定位在一个系统教学内容。

## 这个大纲分成：

- 阶段一： 内容堆集
- 阶段二： 细化二级大纲，甚至 4 级大纲
- 阶段三： 排布具体的课程计划，具体到 1-2 个小时

我们也在招募 Rust 爱好者，一同为中国 Rust 推广助力，如果你有兴趣参与到这个事情，担任讲师， 学习委员，课代表，也可以联系我 Wx: 82565387
- **讲师** 职责 任领课程中的某个教学环节 ，联合学习委员进行程课程输出
- **学习委员** 职责 协助讲师做内容规划或是邀请业界大牛担任讲师， 协助课代表做内容整理
- **课代表** 职责 总结老师讲课内容

你将获得的权益，在本次活动中署名，后期课程内容出品中署名以及 Databend 社区提供的小礼品。

# 阶段一： **内容堆集** 草案

## Rust 入门基本原理

- 讲解变量和值有何不同，以及内存表现
- 所有权机制
- 借用和生命周期

## 类型系统

- 深入了解不同类型在内存中的布局

## 理解宏机制

- 声明性宏
- 过程宏

## FFI

**最后感谢 Databend 社区全力支持**

@joshsulin

**Comment by wubx** [comment]
本次大纲更新中课程不中断， 本周日就可以从第一部分开始了。

**Reply by stardustman** [reply]
个人感觉还是剖析一个一个小例子来讲解比较好。

**Comment by saiksy** [comment]
希望能有更多项目实践经验的分享

**Reply by tanqiangyes** [reply]
希望能有例子，让大家能从应用中学习

**Reply by wubx** [reply]
可以先看看我们前一段的公开课，基本是例子入手。

**Reply by wubx** [reply]
> 

可以考虑直接参与 Databend issue 中：C-good first issue 或是看看对应的 C-good first issue 提交的代码，这个体验也非常不错。

**Comment by wubx** [comment]
优秀内容： https://hashrust.com/blog/moves-copies-and-clones-in-rust/

**Comment by Akagi201** [comment]
希望每次分享时长控制一下, 时间太长很多时候看不完.

**Reply by Akagi201** [reply]
另外非常喜欢极客时间的文字 + 音频形式分享, 方便听音频就听音频, 不方便时候可以直接看文字

**Reply by wubx** [reply]
本次计划是 PDF + 视频  ，音频可能需要自行动手了。

**Comment by wealrain** [comment]
希望能将一些应用层面上的知识，如webassembly方面的，以及rust在云原生方面的一些方向、实例

**Reply by wubx** [reply]
已经邀请了 webassembly 主要 contributor  ， 他马上回国，隔离期间搞起。

**Reply by tanqiangyes** [reply]
@wubx 那也太棒了，最近太需要这部分知识了。

**Comment by hushui502** [comment]
语言层面的东西感觉可以用文章来输出，更希望公开课可以偏向实战/应用一些。

**Reply by wubx** [reply]
这个可能近期会放到周四晚上的内容，可关注周四晚上的公开课。

**Comment by cgfork** [comment]
有没有那种做一个完整工程的课程（可以分为多期），可以看到一些代码设计，风格，工程向的东西

**Reply by wubx** [reply]
本次设计就是希望这个是一个系例，更加紧凑一点。

**Comment by dust1** [comment]
可以分享一些最新的论文研究成果

**Reply by wubx** [reply]
这个可以后面引入新的系例

**Reply by dust1** [reply]
可以偶尔来几门跟XP老师一样的课程，先介绍相关成果的工程实践，然后分享一些最新的研究思路。太多了怕消化不了

**Reply by wubx** [reply]
XP 老师手里还有很多东西，后面会在安排他讲一下。 他讲的风格也很好。

**Comment by Akagi201** [comment]
分享资源: https://github.com/pingcap/talent-plan

**Comment by tanqiangyes** [comment]
老师可以分享下，比如怎么手把手写一个kv存储库，然后将我们课上讲的都用一遍，相信效果会很好。

**Reply by wubx** [reply]
这个需要再往后面放一下，Databend 现在的 KV 使用 Seld 存在有很多问题。我们也在规划决定再写一个了 KV。 如果开干，这个将来也会开源出来， 有兴趣了，可以保持对 https://github.com/datafuselabs 的关注。

**Reply by tanqiangyes** [reply]
可以弄个实验性的，就只是给大家串联下知识结构，不需要涉及到太深的领域知识。

**Reply by dust1** [reply]
简单的KV存储可以看《Designing Data-Intensive Application》这本书来学习

**Reply by tanqiangyes** [reply]
ddia正在啃呢，不过里面的不涉及到rust，只是觉得现在rust基础学了一遍，没串起来，需要一个项目串一下。

**Reply by wubx** [reply]
> 可以弄个实验性的，就只是给大家串联下知识结构，不需要涉及到太深的领域知识。

其实这个可以发起一个组织，组织一些人搞起。

**Reply by tanqiangyes** [reply]
大佬才有人脉啊，我们新手没有的
